### ISSUE TITLE
printonlyreused and capitalizing doesn't work

### ISSUE THREAD

**Issue by trololo01** [oetiker/acronym#25]
Version 1.46 provides options to only print multiply used acronyms ([printonlyreused]) to the list of acronyms as well as capitalizing e.g. long versions (\Acl{XYZ}). Capitalizing worked for me in an earlier version. I updated the package to version 1.46 but none of the above mentioned works now. I also re-installed the required packages suffix and xstring. I'm using MiKTeX 2.9.7380 with PDFlatex and also reset MiKTeX configuration but nothing helped. Please find below a MWE to describe the problem.

In my output, the long version of A is not capitalized, and B is still on the list. 

```
\documentclass[12pt]{article}
\usepackage[printonlyreused]{acronym}
\begin{document}

\section{Chapter}
First use: \ac{NA}  \\
2nd use: \ac{NA}   \\
Long version with capital: \Acl{NA} \\
Other acronym, only mentioned once: \ac{NB} \\
The third acro is not mentioned to be excluded on the printed list

\section{List of acronyms}
\begin{acronym}
\acro{NA}{new acro A}
\acro{NB}{new acro B} % negative test for [printonlyreused]
\acro{NC}{new acro C} % negative test for [printonlyused]
\end{acronym}

\end{document}
```

**Comment by moewew** [comment]
Also asked on TeX.SX https://tex.stackexchange.com/q/534362/35864, where I commented that I suspect https://github.com/oetiker/acronym/commit/c14ea865ed50ffcfea5ebd0de7a0efcf00af09c5 (from https://github.com/oetiker/acronym/pull/21 for #20) might be responsible for the capitalisation issue.

---

I'm not quite sure about the `printonlyreused` issue, but if we introduce some logging into `\acronymused`
```latex
\documentclass[12pt]{article}
\usepackage[printonlyreused]{acronym}

\usepackage{etoolbox}

\makeatletter
\renewcommand*{\acronymused}[1]{%
  \typeout{USED #1}%
  \expandafter\ifx\csname acused@#1@once\endcsname\AC@used%
    \expandafter\ifx\csname acused@#1@twice\endcsname\AC@used%
      \relax
      \typeout{#1 used twice already}%
    \else%
      \typeout{#1 used only once before, setting flag for second use}%
      \global\expandafter\let\csname acused@#1@twice\endcsname\AC@used%
      \global\let\AC@populated\AC@used%
    \fi%
  \else%
    \typeout{#1 not used before}%
    \global\expandafter\let\csname acused@#1@once\endcsname\AC@used%
    \ifAC@printonlyreused%
      \relax%
    \else%
      \global\let\AC@populated\AC@used%
    \fi%
  \fi%
}
\makeatletter

\begin{document}
First use: \ac{NA}
2nd use: \ac{NA}
Long version with capital: \Acl{NA}
Other acronym, only mentioned once: \ac{NB}
The third acro is not mentioned to be excluded on the printed list

\begin{acronym}
\acro{NA}{new acro A}
\acro{NB}{new acro B}
\acro{NC}{new acro C}
\end{acronym}
\end{document}
```
we find
```
(Namenlos-35.aux
USED NA
NA not used before
USED NA
NA used only once before, setting flag for second use
USED NA
NA used twice already
USED NB
NB not used before
)
...
USED NA
NA used twice already
USED NA
NA used twice already
USED NA
NA used twice already
...
(Namenlos-35.aux
USED NA
NA used twice already
USED NA
NA used twice already
USED NA
NA used twice already
USED NB
NB used twice already
)
```
in the `.log`.

Essentially, each call of `\acronymused` is executed three times. The call is written to the `.aux` file, so it is executed once at the beginning of the document, when the `.aux` file is read. It is called again in the document when an acronym is used. This is a problem for the `@twice` counter, since a single use in the `.aux` will have set the `@once` flag, so now we get the `@twice` flag, even though the acronym itself was not used twice. The third call is from reading the `.aux` file back at the end of the document, as far as I can tell that is not particularly problematic, since the list has already been printed at that point.

One way out of this would be not to call `\acronymused` in the document, i.e. to change

https://github.com/oetiker/acronym/blob/c57bc825355e931198fbe0f064e9a6216b1c106c/acronym.dtx#L1042-L1046

to
```latex
\renewcommand*{\AC@logged}[1]{%
   \@bsphack
   \protected@write\@auxout{}{\string\acronymused{#1}}%
   \@esphack}
```

but that would mean that `acronym` can not detect if an acronym was used directly on the first run. It would always require another LaTeX run with `.aux` file to check if the acronym was used or not. In practice that means that the list of acronyms is empty when all the acronyms show "**NA!**" and ```Package acronym Warning: Acronym `NA' is not defined on input line 31.```.

**Comment by moewew** [comment]
Some more investigation about the capitalisation issue.

The problem is indeed https://github.com/oetiker/acronym/commit/c14ea865ed50ffcfea5ebd0de7a0efcf00af09c5 (#21).

If we redefine `\@firstupper` for a moment to see what that macro sees
```latex
\documentclass[12pt]{article}
\usepackage{acronym}

\makeatletter
\renewcommand{\@firstupper}[1]{%
    \texttt{|\detokenize{#1}|}%
}
\makeatother


\begin{document}

\section{Chapter}
Long version with capital: \Acl{NA}
\section{List of acronyms}
\begin{acronym}
\acro{NA}{new acro A}
\end{acronym}
\end{document}
```
we find that `\@firstupper` gets only fed `\expandafter` no matter what the acronym actually is. The real acronym is never passed to `\@firstupper`.

If one wanted to expand `\AC@aclp` before it got to `\@firstupper` the correct code would be
```latex
\newcommand*\AC@Aclp[1]{%
  \expandafter\@firstupper\expandafter{\AC@aclp{#1}}%
}
```
instead of
https://github.com/oetiker/acronym/blob/c57bc825355e931198fbe0f064e9a6216b1c106c/acronym.dtx#L1441-L1443

In this case, however, there is pretty much no difference between `\expandafter\@firstupper\expandafter{\AC@aclp{#1}}%` and
```latex
\newcommand*\AC@Aclp[1]{%
  \@firstupper{\AC@aclp{#1}}%
}
```
due to the way `\@firstupper` is defined. That brings us nicely to the core of the problem here.

`\@firstupper` is defined using `xstring` commands and `xstring` performs full `\edef` expansion on its argument to make sure it works with a string and not just with macros containing (or rather eventually expanding to) strings. But not everything in TeX is expandable in an `\edef`, in particular code for messages as produced by `\PackageWarning` is not expandable and will instead produce (often not really helpful) error messages. This problem only occurs if no `.aux` file is present, since `\AC@get` produces a warning if the acronym isn't found in the `.aux` and just expands to the acronym otherwise

https://github.com/oetiker/acronym/blob/c57bc825355e931198fbe0f064e9a6216b1c106c/acronym.dtx#L1478-L1484

We can see roughly what `\@firstupper` tries to uppercase in that case in the following MWE (it's not quite accurate, but should be good enough to understand the issue)
```latex
\documentclass[12pt]{article}
\usepackage{acronym}

\makeatletter
\renewcommand*\AC@Acl[1]{%
  \@firstupper{\AC@acl{#1}}%
}
\renewcommand*\AC@Aclp[1]{%
  \@firstupper{\AC@aclp{#1}}%
}
\renewcommand{\@firstupper}[1]{%
  \expandafter\def\expandafter\mytempa\expandafter{\romannumeral0#1}%
  \texttt{|\expandafter\detokenize\expandafter{\mytempa}|}%
}
\makeatother


\begin{document}

\section{Chapter}
Long version with capital: \Acl{NA}
\section{List of acronyms}
\begin{acronym}
\acro{NA}{new acro A}
\end{acronym}
\end{document}
```
On a run without `.aux` that produces
>```latex
>|\protect \GenericWarning {(acronym)\@spaces \@spaces \@spaces \@spaces }{Package acronym Warning: Acronym ‘NA’ is not defined}\textbf {NA!}\else \expandafter \@secondoftwo \fn@NA \fi |
>```
for me. That is not something the normal definition of `\@firstupper` can deal with, so it breaks.

The following would work
```latex
\documentclass[12pt]{article}
\usepackage{acronym}

\makeatletter
\renewcommand*\AC@Acl[1]{%
  \@firstupper{\AC@acl{#1}}%
}
\renewcommand*\AC@Aclp[1]{%
  \@firstupper{\AC@aclp{#1}}%
}
\renewcommand*\AC@get[3]{%
    \ifx#1\relax
       sorry, no acronym, but expandable%
    \else
       \expandafter#2#1%
    \fi}
\makeatother


\begin{document}

\section{Chapter}
Long version with capital: \Acl{NA}
\section{List of acronyms}
\begin{acronym}
\acro{NA}{new acro A}
\end{acronym}
\end{document}
```
but of course does not produce the desired warning on the first run.

Unfortunately, I don't have an idea for a quick solution here. As things are currently, you would have to check first if the acronym exists and then only after you have made sure apply `\@firstupper`. But that still requires some rewriting of the code.

**Comment by oetiker** [comment]
could we somehow set a flag 'outside' and use @firstupper inside if the flag is set ?

**Comment by marsmee** [comment]
The capitalisation bug should be fixed with #26 now. Unfortunately, there seems to be no proper way for `xstring` to handle our code. Using `\expandarg` or `\normalexpandarg` also doesn't have the desired effect.

I will try to fix the other issue soon. Nevertheless, the feature seems to work as intended when the list of acronyms appears at the beginning of the document. This was actually my test case. Anyway, it should work properly everywhere.
